pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexportendpointslice
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+)
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+
+	// TODO (fgiloux): When the information is available in the cache server
+	// check if at least one APIBinding is bound in the shard to the APIExport referenced by the APIExportEndpointSLice.
+	// If so, add the respective endpoint to the status.
+	// For now the unfiltered list is added.
+
+	// Check for valid reference
+	workspaceRef := apiExportEndpointSlice.Spec.APIExport.Workspace
+	if workspaceRef == nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Missing APIExport reference",
+		)
+		return nil
+	}
+
+	// Get APIExport
+	apiExportClusterName, err := getAPIExportClusterName(apiExportEndpointSlice)
+	if err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return nil
+	}
+
+	apiExport, err := c.getAPIExport(apiExportClusterName, workspaceRef.ExportName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Don't keep the endpoints if the APIExport has been deleted
+			apiExportEndpointSlice.Status.APIExportEndpoints = nil
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.APIExportNotFoundReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"APIExport %s|%s not found",
```

Comment:
Does this leak information? By creating APIExportSlices, I get to figure out which exports exist and where?

Reply:
I would think an APIExportEndpointSlice may only point at APIExports in the same Workspace. Is that accurate? Maybe we need to adjust APIExportEndpointSliceSpec slightly so it can't point at an APIExport in a different Workspace?

Reply:
> I would think an APIExportEndpointSlice may only point at APIExports in the same Workspace. Is that accurate? 

APIExportEndpointSlice is similar to an APIBinding and can point to an APIExport in another workspace. It is desired so. Service provider wanting to cover shards in multiple regions, will want more often than not an APIExportEndpointSlice in each of them with the relevant subset of endpoints. Partitions will allow that.

> By creating APIExportSlices, I get to figure out which exports exist and where?

We would probably need some mechanism similar to what is available for APIBinding where you need privileges to bind an APIExport. I don't know the implementation details of it. It could be something we create an issue for and follow up with a subsequent PR. What do you think?

Reply:
Hm, what do we expose on APIBinding failing to find an export? I would be surprised if it's the raw error. Let's not add information leaks :)

Reply:
Guess where I copied the code from :)
https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/apis/apibinding/apibinding_reconcile.go#L105-L131

Seriously, I see your point and will remove the raw error. I could do the same for APIBinding in a subsequent PR.